File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/consumer/aop/ConsumeMqMessage.java
package org.xyattic.eventual.consistency.support.core.consumer.aop;

import java.io.Serializable;
import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Inherited;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * @author wangxing
 */
@Target({ElementType.METHOD, ElementType.ANNOTATION_TYPE})
@Retention(RetentionPolicy.RUNTIME)
@Documented
@Inherited
public @interface ConsumeMqMessage {

    String persistenceName() default "";

    String transactionManager() default "";

    Class<? extends Serializable> messageClass() default DefaultMessageClass.class;

    String messageExpression() default "";

    String messageProvider() default "";

    String messageIdExpression();

    class DefaultMessageClass implements Serializable {
    }

}

File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/consumer/aop/ConsumeMqMessage.kt
package org.xyattic.eventual.consistency.support.core.consumer.aop

import org.xyattic.eventual.consistency.support.core.consumer.AbstractMessage
import java.lang.annotation.Inherited
import kotlin.reflect.KClass

/**
 * @author wangxing
 */
@Target(
    AnnotationTarget.FUNCTION,
    AnnotationTarget.PROPERTY_GETTER,
    AnnotationTarget.PROPERTY_SETTER,
    AnnotationTarget.ANNOTATION_CLASS
)
@Retention(AnnotationRetention.RUNTIME)
@MustBeDocumented
@Inherited
annotation class ConsumeMqMessage(
    val persistenceName: String = "",
    val transactionManager: String = "",
    val messageClass: KClass<out AbstractMessage> = DefaultMessageClass::class,
    val messageExpression: String = "",
    val messageProvider: String = "",
    val messageIdExpression: String = ""
)

internal class DefaultMessageClass : AbstractMessage() {
    override fun getId(): String {
        TODO("Not yet implemented")
    }
}
